core/src/program/blocks/span_block.rs

```diff
@@ -335,7 +338,20 @@ fn batch_ops(ops: Vec<Operation>) -> (Vec<OpBatch>, Digest) {
         batches.push(batch);
     }
 
-    let hash = hasher::hash_elements(flatten_slice_elements(&batch_groups));
+    // compute total number of operation groups in all batches. This is done as follows:
+    // - For all batches but the last one we set the number of groups to 8, regardless of the
+    //   actual number of groups in the batch. The reason for this is that when operation
+    //   batches are concatenated together each batch contributes 8 elements to the hash.
+    // - For the last batch, we take the number of actual batches and round it up to the next
+    //   power of two. The reason for rounding is that the VM always executes a number of
+    //   operation groups which is a power of two.
+    let num_batches = batches.len();
+    let last_batch_num_groups = batches[num_batches - 1].num_groups().next_power_of_two();
+    let num_op_groups = (num_batches - 1) * BATCH_SIZE + last_batch_num_groups;
```

Comment:
One thought - this logic comes up in a few different places (including in the next PR with the aux hints and `get_span_op_group_count` in this PR). It might make things clearer if there's a way to use one well-documented helper that gets called in the necessary spots (or a couple of related helpers that live together). The logic would only need to be explained in one place, and the calls would link the related sections of the code together. (I know there's a good chance you've already considered this and there might not be a straightforward way to do it cleanly)

Reply:
Good point! Fixed.